### PR TITLE
Start a job using Executor K8s

### PR DESCRIPTION
**Pull request by FenrirUnbound**
When the Builds plugin is asked to start a build, it will save some initial build data and then begin the job using the `screwdriver-executor-k8s` module.

Fun things to point out:
* The job model contains all the business logic. We should consider moving it into an external module so that other plugins can interact with the data.
* It was fun 😺 

Blocked by screwdriver-cd/executor-k8s#1

### CODE REVIEW

File: lib/buildmodel.js [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+            number: Date.now()
+        });
+
+        async.waterfall([
+            async.apply(this.datastore.save, {
+                table: 'builds',
+                params: {
+                    id,
+                    data: {
+                        jobId,
```
Comment (nkatzman) [inline lib/buildmodel.js]: I think we should add defaults for all of the data expected here for the build model

Comment (nkatzman) [inline lib/buildmodel.js]: https://github.com/screwdriver-cd/data-model/blob/master/model/build.js#L98

File: lib/buildmodel.js [hunk 2]
```diff
@@ -0,0 +1,100 @@
+'use strict';
+const async = require('async');
+const Executor = require('screwdriver-executor-k8s');
+const hashr = require('screwdriver-hashr');
+
+/**
+ * Given a Job ID, look up the associatd Pipeline ID
```
Comment (d2lam) [inline lib/buildmodel.js]: `associated`

File: lib/buildmodel.js [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+ *
+ * TODO: Use JobModel object instead
+ * @method fetchPipelineId
+ * @param  {Object}        datastore The datastore object that can retrieve the data from the datasore
+ * @param  {String}        jobId     The ID of the job to look at
+ * @param  {Function}      callback  Callback function
```
Comment (stjohnjohnson) [inline lib/buildmodel.js]: What's the callback function?

File: lib/buildmodel.js [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+            pipelineId,
+            scmUrl: data.scmUrl
+        });
+    });
+}
+
+const BuildModel = class {
+    constructor(datastore) {
+        this.datastore = datastore;
+        this.executor = new Executor({});
```
Comment (stjohnjohnson) [inline lib/buildmodel.js]: This should probably be passed in from the plugin options

File: lib/buildmodel.js [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+ * Given a Pipeline ID, look up the associated SCM Url
+ *
+ * TODO: Use PipelineModel instead
+ * @method fetchScmUrl
+ * @param  {Object}        datastore The datastore object that can retrieve the data from the datasore
+ * @param  {String}    pipelineId    The ID of the pipeline to look at
+ * @param  {Function}  callback      fn(err, result), where result is an object consisting of two keys: "pipelineID"
+ *                                   and "scmUrl"
+ */
+function fetchScmUrl(datastore, pipelineData, callback) {
```
Comment (d2lam) [inline lib/buildmodel.js]: the 2nd parameter is `jobData`, not `pipelineData` right?

**Comment by FenrirUnbound** [comment]
Waiting on screwdriver-cd/executor-k8s to have something that will not crash the system

File: lib/buildmodel.js (hunk 1 above)
Comment (FenrirUnbound) [inline lib/buildmodel.js]: 👍 Addressed 😸

**Comment by FenrirUnbound** [comment]
PR at screwdriver-cd/executor-k8s#1 has been merged 🎆

File: lib/create.js [hunk 6]
```diff
@@ -1,50 +1,41 @@
...
 
-            datastore.save(config, (err) => {
+            build.create(request.payload, (err, data) => {
                 if (err) {
                     return reply(boom.wrap(err));
                 }
 
-                const result = hoek.merge({ id: newId },
+                const buildId = data.id;
+                const result = hoek.merge({ id: buildId },
```
Comment (nkatzman) [inline lib/create.js]: should the result contain the entire data object that was created?

Comment (FenrirUnbound) [inline lib/create.js]: Passing the entire object back is desirable. Basing how Github API behaves, this is similar to how their POST requests respond

Comment (nkatzman) [inline lib/create.js]: They only pass back the data that was requested in POST? We'd have to GET the data if we wanted to get what was actually created?

Comment (d2lam) [inline lib/create.js]: or we can modify the create to return the whole object that got created instead of just the id?

Comment (FenrirUnbound) [inline lib/create.js]: Ideally the model would. We're trying to do a minimal model here to see what the real deal looks like.

For the week, we're trying for the goal. The real model will need to be worked on after we hit our goal; a gold bar to the week's work :smiley_cat:

File: lib/buildmodel.js [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+     * @param  {String}   config.container The kind of container to use
+     * @param  {Function} callback         Callback function
+     */
+    create(config, callback) {
+        const container = config.container;
+        const jobId = config.jobId;
+        const now = Date.now();
+        const id = hashr.sha1({
+            jobId,
+            number: now
```
Comment (nkatzman) [inline lib/buildmodel.js]: should this be `runNumber: now` ? Not like it _really_ matter, but just keep things as values in the buildData

File: lib/buildmodel.js [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+const Executor = require('screwdriver-executor-k8s');
+const hashr = require('screwdriver-hashr');
+const hoek = require('hoek');
+
+/**
+ * Given a Job ID, look up the associated Pipeline ID
+ *
+ * TODO: Use JobModel object instead
+ * @method fetchPipelineId
+ * @param  {Object}        datastore The datastore object that can retrieve the data from the datasore
```
Comment (tkyi) [inline lib/buildmodel.js]: `datasore`

File: lib/create.js (hunk 6 above)
Comment (nkatzman) [inline lib/create.js]: result can just be the data returned by `build.create`

File: lib/buildmodel.js [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+        async.waterfall([
+            async.apply(this.datastore.save, {
+                table: 'builds',
+                params: {
+                    id,
+                    data: initialBuildData
+                }
+            }),
+            (next) => {
+                fetchPipelineId(this.datastore, jobId, next);
```
Comment (nkatzman) [inline lib/buildmodel.js]: Would doing `async.apply(this.datastore, jobId)` make this a little cleaner?

File: package.json [hunk 10]
```diff
@@ -41,10 +41,12 @@
     "sinon": "^1.17.4"
   },
   "dependencies": {
+    "async": "^2.0.0-rc.6",
     "boom": "^3.2.1",
     "hoek": "^4.0.1",
     "joi": "^8.1.1",
     "screwdriver-data-model": "^2.0.0",
+    "screwdriver-executor-k8s": "0.0.1",
```
Comment (tkyi) [inline package.json]: should be `0.0.2`

**Comment by tkyi** [comment]
For future work, might be cleaner to move buildmodel.js into another folder of `lib/models`. Otherwise, looks good once `package.json` is updated.

**Comment by nkatzman** [comment]
👍

**Comment by tkyi** [comment]
👍

**Comment by d2lam** [comment]
👍

**Pull request merged**